TrackBuffer.decodeQueue() shouldn't be a public method.

jyavenard opened the pull request:
#### 0c01b830e20090c83cef7692e6236ff4b1171ed2
<pre>
TrackBuffer.decodeQueue() shouldn&apos;t be a public method.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293459">https://bugs.webkit.org/show_bug.cgi?id=293459</a>
<a href="https://rdar.apple.com/151887171">rdar://151887171</a>

Reviewed by Youenn Fablet.

We move the whole logic to add a new sample to the TrackBuffer in the TrackBuffer class instead.
Similarly, we move the logic to retrieve the next frame to be enqueued for decoding in the TrackBuffer.

This allows for some simplification and removal of duplicated code.
It will also allows to make the tracking of the next minimumUpcomingSampleTime
dynamic required for a follow-up change.

* Source/WebCore/platform/graphics/TrackBuffer.cpp:
(WebCore::TrackBuffer::addSample): addSample now handles adding the sample to the global sample map and the decode queue.
(WebCore::TrackBuffer::nextSample): Method added.

Canonical link: <a href="https://commits.webkit.org/295339@main">https://commits.webkit.org/295339@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1966e0b7b2a2a75b4eb5b8ace07f27ce722b794b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24505 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110006 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55466 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106832 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33050 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79570 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107798 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19360 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94573 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59877 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12650 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54848 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88815 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12697 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112413 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31957 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23496 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88649 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32321 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90799 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88276 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22500 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33166 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10930 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27270 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31882 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37236 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31674 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35015 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33233 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -301,43 +301,26 @@ void SourceBufferPrivate::provideMediaData(TrackBuffer& trackBuffer, TrackID tra
+            ++enqueuedSamples;
 #endif
+        } else
+            break;
```
I would tend to write if as `if (!sample) break;  // proceed with enqueuing`

jyavenard replied:
agree. I hesitated a bit on it.

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1120,38 +1120,25 @@ - (void)removeAllVideoTargets;
+        if (auto sample = trackBuffer.nextSample()) {
+            enqueueSample(sample.releaseNonNull(), trackId);
+            ++enqueuedSamples;
+        } else
```
Ditto here

youennf reviewed (approved):
nice

jyavenard committed:
TrackBuffer.decodeQueue() shouldn't be a public method.
https://bugs.webkit.org/show_bug.cgi?id=293459
rdar://151887171

Reviewed by Youenn Fablet.

We move the whole logic to add a new sample to the TrackBuffer in the TrackBuffer class instead.
Similarly, we move the logic to retrieve the next frame to be enqueued for decoding in the TrackBuffer.

This allows for some simplification and removal of duplicated code.
It will also allows to make the tracking of the next minimumUpcomingSampleTime
dynamic required for a follow-up change.

* Source/WebCore/platform/graphics/TrackBuffer.cpp:
(WebCore::TrackBuffer::addSample): addSample now handles adding the sample to the global sample map and the decode queue.
(WebCore::TrackBuffer::nextSample): Method added.

Canonical link: https://commits.webkit.org/295339@main
